text/0000-nested-publish.md

```diff
@@ -114,36 +115,59 @@ Then you can `cargo publish` from within the parent directory `foo/`, and this w
-
-The presence or absence of a `[workspace]` has no effect on the new behavior, just as it has no effect on existing package publication. Nested packages may use workspace inheritance.
+## Cargo: Manifest
+
+Two new possible values are added to the manifest.
+
+* The `package.publish` field allows `"nested"` as a value, in addition to existing `false` and `true`. This value affects `cargo publish` and nested dependencies as described below.
+* The `dependencies.*.publish` field is newly defined, with the only currently allowed value being `"nested"`, to declare that that dependency is a nested dependency. (If desired, `publish = false` could be also be permitted to explicitly document an intent not to nest; this would be identical to the status quo and to omitting the key.)
+    * It is an error if a nested dependency does not have a `path` field, or if it has a `version`, `git`, or any other package source field, unless future work defines a meaning for that combination.
+
+When a nested dependency is present (making its referent be a nested package), the following additional requirements apply:
+
+* The nested package must have `package.publish = "nested"`; both `false` and `true` are errors.
+
+* If the nested package specifies `package.license`, the parent package's (not necessarily the dependent's) license expression must comply with the nested package's. This check is done solely in terms of the operators in the license expression. For example, if two nested packages contain licenses of `MIT` and `BSD-3-Clause`, then the parent package's expression must be `MIT AND BSD-3-Clause` or similar.
+
+  This check is intended only to prevent accidents (such as vendoring a third-party package without considering the implications of redistributing it). It is always valid to omit `package.license` from the nested package, thus making no machine-readable claims about its licensing independent from the parent package's.
+
+* We might want to explicitly prohibit nested packages from specifying a `package.version`, to avoid giving the misleading impression that it means anything. (`package.version` is already optional as of Cargo 1.75, but this is currently equivalent to `version = "0.0.0"`.)
+
+## **`cargo package` &amp; `cargo publish`**
+
+If `cargo package` or `cargo package` is directed to operate on a package whose manifest specifies `package.publish = "nested"`, that is an error. (Users wishing to make crates in a nested package public should either write `package.publish = true` or should re-export their contents from another package.)
+
+When a valid parent package is packaged, each of its transitive nested dependencies must be included in the `.crate` archive file. This has two sub-cases:
+
+* The nested package may be in a subdirectory of the parent package directory. In this case, it is copied to the same location in the archive.
+* Otherwise, it is copied to `.cargo/packages/<package name>/` within the archive.
```

Comment:
Can there be duplicate package names?  What happens in that case?

Reply:
I was thinking that duplicate package names would be prohibited within a given parent package, as they currently are for local `path` dependencies. I should add that to the text.

Reply:
I have specified that package names must be unique within a parent package; and furthermore that this is an error at build time and not just packaging time (so that the lockfile need not deal with distinguishing different arbitrarily-deep duplicates).